File: platforms/ios/VeilTests/BroadcastMessageTests.swift
// Copyright (c) 2018, salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause

import XCTest
import WebKit

@testable import Veil

class BroadcastMessageTests: XCTestCase, WKNavigationDelegate {
    let bridge = ConnectionTestBridge()
    var webView: WKWebView!
    var loadingExpectation: XCTestExpectation?

    override func setUp() {
        let userContentController = WKUserContentController()
        let appScript = Bundle(for: EvalJsTests.self).url(forResource: "broadcastJs", withExtension: "js")
            .flatMap { try? NSString(contentsOf: $0, encoding: String.Encoding.utf8.rawValue ) }
            .flatMap { WKUserScript(source: $0 as String, injectionTime: .atDocumentEnd, forMainFrameOnly: true) }
        if let appScript = appScript {
            userContentController.addUserScript(appScript)
        }
        let config = WKWebViewConfiguration()
        config.userContentController = userContentController
        self.webView = WKWebView(frame: CGRect.zero, configuration: config)
        self.webView.navigationDelegate = self
        let _ = Connection(from: webView, to: bridge, as: "ConnectionTestBridge")
    }

    override func tearDown() {
        self.webView.navigationDelegate = nil
        self.webView = nil
    }

    func loadWebViewAndWait(html: String = "<html><body></body></html>") {
        loadingExpectation = expectation(description: "web view loaded")
        self.webView.loadHTMLString(html, baseURL: nil)
        wait(for: [loadingExpectation!], timeout: 10)
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        loadingExpectation?.fulfill()
    }

    func testBroadcastMessageWithNoParam() {
        loadWebViewAndWait()
        let x = expectation(description: "js result")
        var callCountMatches = false
        self.webView.broadcastMessage(name: "testMessageWithNoParam") { (calledHandlerCount, error) -> () in
            if 1 == calledHandlerCount {
                callCountMatches = true
            }
            x.fulfill()
        }
        wait(for: [x], timeout: 5)
        XCTAssertTrue(callCountMatches)
    }

    func testBroadcastMessageWithParam() {
        loadWebViewAndWait()
        let x = expectation(description: "js result")
        var callCountMatches = false
        let testArg = UserDefinedType()
        self.webView.broadcastMessage(name: "testMessageWithParam", arg: testArg) { (calledHandlerCount, error) -> () in
            if 1 == calledHandlerCount {
                callCountMatches = true
            }
            x.fulfill()
        }
        wait(for: [x], timeout: 5)
        XCTAssertTrue(callCountMatches)
    }

    func testBroadcastMessageWithNoHandler() {
        loadWebViewAndWait()
        // Test unsubscribing of a message handler
        let x = expectation(description: "js result")
        var callCountMatches = false
        self.webView.broadcastMessage(name: "testUnsubscribingHandler") { (calledHandlerCount, error) -> () in
            if 1 == calledHandlerCount {
                callCountMatches = true
            }
            x.fulfill()
        }
        wait(for: [x], timeout: 5)
        XCTAssertTrue(callCountMatches)

        // Test that no message handler is called
        let y = expectation(description: "js result")
        callCountMatches = false
        self.webView.broadcastMessage(name: "testMessageWithNoParam") { (calledHandlerCount, error) -> () in
            if 0 == calledHandlerCount {
                callCountMatches = true
            }
            y.fulfill()
        }
        wait(for: [y], timeout: 5)
        XCTAssertTrue(callCountMatches)
    }
}

File: platforms/android/demo-app/src/androidTest/java/com/salesforce/veildemoapp/BroadcastMessageTests.kt
// Copyright (c) 2019, salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause


package com.salesforce.veildemoapp

import android.support.test.internal.runner.junit4.statement.UiThreadStatement.runOnUiThread
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.util.Log
import android.webkit.WebView
import android.webkit.WebViewClient
import com.salesforce.veil.Connection
import com.salesforce.veil.ResourceUtils
import com.salesforce.veil.addConnection
import com.salesforce.veil.broadcastMessage
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@RunWith(AndroidJUnit4::class)
class BroadcastMessageTests {
    init {
        WebView.setWebContentsDebuggingEnabled(true);
    }

    @Rule
    @JvmField
    val activityRule: ActivityTestRule<WebViewActivity> = ActivityTestRule<WebViewActivity>(WebViewActivity::class.java, false, true)

    var webView: WebView? = null

    @Before
    fun setup() {
        webView = activityRule.activity.webView
        val latch = CountDownLatch(1)
        runOnUiThread {
            webView?.let {
                val bridge = ConnectionTests.TestBridge()
                val veilPreScript = ResourceUtils(it.context).stringFromRawResource(R.raw.broadcasttests)
                it.addConnection(bridge, "TestBridge", veilPreScript)
            }
        }
        latch.await(5, TimeUnit.SECONDS)
        runOnUiThread {
            webView?.let {
                it.loadDataWithBaseURL(null, "<html><body></body></html>", "text/html", "UTF-8", null)
            }

        }
        latch.await(5, TimeUnit.SECONDS)
    }

    @Test
    fun broadcastMessageWithNoParam() {
        var callCountMatches = false
        val latch = CountDownLatch(1)
        webView?.let {
            runOnUiThread {
                it.broadcastMessage("testMessageWithNoParam") {
                    if (1 == it) {
                        callCountMatches = true
                        latch.countDown()
                    }
                }
            }
        }
        latch.await(5, TimeUnit.SECONDS)
        assertEquals(true, callCountMatches)
    }

    @Test
    fun broadcastMessageWithParam() {
        var callCountMatches = false
        val latch = CountDownLatch(1)
        webView?.let {
            runOnUiThread {
                val testArg = MainActivity.UserDefinedType()
                it.broadcastMessage("testMessageWithParam", testArg) {
                    if (1 == it) {
                        callCountMatches = true
                        latch.countDown()
                    }
                }
            }
        }
        latch.await(5, TimeUnit.SECONDS)
        assertEquals(true, callCountMatches)
    }

    @Test
    fun broadcastMessageWithNoHandler() {
        // Test unsubscribing of a message handler
        var callCountMatches = false
        var latch = CountDownLatch(1)
        webView?.let {
            runOnUiThread {
                it.broadcastMessage("testUnsubscribingHandler") {
                    if (1 == it) {
                        callCountMatches = true
                        latch.countDown()
                    }
                }
            }

        }
        latch.await(5, TimeUnit.SECONDS)
        assertEquals(true, callCountMatches)

        // Test that no message handler is called
        callCountMatches = false
        latch = CountDownLatch(1)
        webView?.let {
            runOnUiThread {
                it.broadcastMessage("testMessageWithNoParam") {
                    if (0 == it) {
                        callCountMatches = true
                        latch.countDown()
                    }
                }
            }
        }
        latch.await(5, TimeUnit.SECONDS)
        assertEquals(true, callCountMatches)
    }
}
